src/backend/distributed/connection/connection_management.c

```diff
@@ -527,19 +535,207 @@ FindAvailableConnection(dlist_head *connections, uint32 flags)
 
 
 /*
- * ErrorIfMultipleMetadataConnectionExists throws an error if the
- * input connection dlist contains more than one metadata connections.
+ * EnsureTxStateAndGetHealthyConnections is a helper function that goes
+ * over the input connections, and:
+ *  - Errors for the connections whose remote transaction is broken
+ *    when checkTransactionState=true
+ *  - Skips connections that are not healthy, marks those to be closed
+ *    at the end of the transaction such that they do not linger around
+ *
+ * And, finally returns the list (unless already errored out) of connections
+ * that can are healthy and can be reused.
  */
-static void
-ErrorIfMultipleMetadataConnectionExists(dlist_head *connections)
+static List *
+EnsureTxStateAndGetHealthyConnections(dlist_head *connections,
+									  bool checkTransactionState)
 {
-	bool foundMetadataConnection = false;
+	List *healthyConnections = NIL;
+
 	dlist_iter iter;
 	dlist_foreach(iter, connections)
 	{
 		MultiConnection *connection =
 			dlist_container(MultiConnection, connectionNode, iter.cur);
 
+		RemoteTransaction *transaction = &connection->remoteTransaction;
+		bool inRemoteTrasaction =
+			transaction->transactionState != REMOTE_TRANS_NOT_STARTED;
+
+		if (checkTransactionState && inRemoteTrasaction)
+		{
+			PGTransactionStatusType status = PQtransactionStatus(connection->pgConn);
+
+			/* if the connection is in a bad state, so is the transaction's state */
+			if (status == PQTRANS_INERROR || status == PQTRANS_UNKNOWN)
+			{
+				transaction->transactionFailed = true;
+			}
+
+			if (transaction->transactionFailed &&
+				(transaction->transactionCritical ||
+				 !dlist_is_empty(&connection->referencedPlacements)))
+			{
+				/*
+				 * If a connection is executing a critical transaction or accessed any
+				 * placements, we should not continue the execution.
+				 */
+				ReportConnectionError(connection, ERROR);
+			}
+		}
+
+		if (connection->pgConn == NULL || PQsocket(connection->pgConn) == -1 ||
+			PQstatus(connection->pgConn) == CONNECTION_BAD ||
+			connection->connectionState == MULTI_CONNECTION_FAILED ||
+			RemoteSocketClosed(connection))
```

Comment:
would it be possible/practical to check all connections in one WaitEvenSet? we can easily have ~16 connections in the list

Reply:
Sure, that is possible. However, with this current approach -- where we check not only for socket close but also for connection/tx state -- the code turned out to be relatively complex. 

So, I can see 3 possibilities here:

a) Leave as-is, given there is no observed perf overhead. Add a comment, suggesting this can be done for all connections at once

b) Drop connection/tx state checks, only focus on RemoteSocketClosed. Then, a single WaitEventSet is relatively easy to implement

c) Keep connection/tx state checks and also use a single waiteventset. This makes code harder to follow. The reason is that we should somehow extract the `socketClosed` information per connection. 
c.1) We can put this as a field to MultiConnection c.2) return a list of connections whose sockets are closed.


I have a slight preference for (a), but I'm fine with (c1). Do you have any preference / other suggestions?

Reply:
To be honest, it bugs me that we're doing all this stuff outside of the executor where we spent a few years tweaking things like socket changes, dealing with multiple connections, doing the right things with latches, connection cache management, handling failures, etc. To me, the right way to solve this is still connection retries in the executor.

Anyway, my suggestion would be to make a WaitEventSet pass on all the sockets and call MarkRemoteTransactionFailed for those sockets that are closed, though that does not exempt us from doing the right thing, e.g. using CHECK_FOR_INTERRUPTS when the latch is set, or checking for -1 sockets.

Reply:
Sketch of an alternative solution: https://github.com/citusdata/citus/pull/6283